pkg/ebpf/probes/probes.go

```diff
@@ -2,178 +2,503 @@ package probes
+		switch a := arg.(type) {
+		case *bpf.Module:
+			module = a
+		case *net.Interface:
+			netIface = a
+		}
+	}
+
+	if attach && module == nil {
+		return fmt.Errorf("incorrect arguments for program: %s", p.programName)
+	}
+
+	if !attach {
+		if netIface == nil { // detach from all if no interface given
+			return p.detachTcAll(module)
+		}
+
+		return p.detachTc(module, netIface)
+	}
+
+	return p.attachTc(module, netIface)
+}
+
+// attachTc attaches a tc program to a given interface
+func (p *tcProbe) attachTc(module *bpf.Module, netIface *net.Interface) error {
+	var err error
+
+	// loopback devices are special, some tc probes should be skipped
+	isNetIfaceLo := netIface.Flags&net.FlagLoopback == net.FlagLoopback
+	if isNetIfaceLo && p.skipLoopback {
 		return nil
 	}
 
-	err := p.link.Destroy()
+	if _, ok := p.tcHooks[netIface]; ok {
+		return fmt.Errorf("%s already attached to %s", p.programName, netIface.Name)
+	}
+
+	hook := module.TcHookInit()
+	if hook == nil {
+		return fmt.Errorf("could not initialize TcHook %s", p.programName)
+	}
+
+	err = hook.SetInterfaceByName(netIface.Name)
+	if err != nil {
+		return fmt.Errorf("failed to set tc hook interface: %v", err)
+	}
+
+	hook.SetAttachPoint(p.tcAttachPoint)
+
+	err = hook.Create()
+	if err != nil {
+		if errno, ok := err.(syscall.Errno); ok && errno != syscall.EEXIST {
+			return fmt.Errorf("tc hook create: %v", err)
+		}
+	}
+
+	prog, _ := module.GetProgram(p.programName)
+	if prog == nil {
+		return fmt.Errorf("could not find tc program %s", p.programName)
+	}
+
+	tcOpts := bpf.TcOpts{ProgFd: int(prog.GetFd())}
 
+	err = hook.Attach(&tcOpts)
 	if err != nil {
-		return fmt.Errorf("error detaching probe %s: %v", p.event, err)
+		return fmt.Errorf("tc attach: %v", err)
+	}
+
+	if p.tcHooks == nil {
+		p.tcHooks = make(map[*net.Interface]*bpf.TcHook)
+	}
+
+	p.tcHooks[netIface] = hook
+
+	return err
+}
+
+// detachTc detaches a tc program from a given interface it is attached to
+func (p *tcProbe) detachTc(module *bpf.Module, netIface *net.Interface) error {
+	var err error
+
+	if _, ok := p.tcHooks[netIface]; !ok {
+		return fmt.Errorf("%s not attached to %s", p.programName, netIface.Name)
+	}
+
+	hook := p.tcHooks[netIface]
+
+	// TODO: https://github.com/aquasecurity/tracee/issues/1828
+	//
+	// 1. Entire clsact qdisc is purged when tc hook is destroyed:
+	//    it might destroy existing filters (other eBPF programs)
+	//    unrelated to tracee.
+	//
+	// 2. eBPF TC detach does not work (destroy needs to be used)
+
+	if hook != nil {
+		err = hook.Destroy()
+		if err != nil {
+			return err
+		}
+	}
+
+	delete(p.tcHooks, netIface)
+
+	return err
+}
+
+// detachTcAll detaches a tc program from all interfaces it is attached to
+func (p *tcProbe) detachTcAll(module *bpf.Module) error {
+	var err error
+
+	for netIface := range p.tcHooks {
+		err = p.detachTc(module, netIface)
+		if err != nil {
+			return err
+		}
 	}
 
-	p.link = nil
 	return err
 }
 
-type Handle int32
```

Comment:
Not sure why the name change to EventHandle. 
Asking because the package explicitly removes itself from the events, so this was a deliberate naming choice (as in probeHandle -> probes.Handle).

Reply:
Yes, I wonder the same

Reply:
![image](https://user-images.githubusercontent.com/7395852/175932404-a092d040-2498-492d-98de-ad3c7a78844f.png)